File: workbook/ch08/data/pqueue.c
#include <stdio.h>
#include <stdlib.h>

typedef struct {
    int priority;
    int value;
} PQNode;

typedef struct {
    PQNode *heap;
    int size;
    int capacity;
    int isMinHeap;  // 1 for Min-Heap, 0 for Max-Heap
} PriorityQueue;

PriorityQueue* createPQ(int capacity, int isMinHeap);
void insertPQ(PriorityQueue *pq, int priority, int value);
int extractPQ(PriorityQueue *pq);
void heapifyUp(PriorityQueue *pq, int index);
void heapifyDown(PriorityQueue *pq, int index);
void swap(PQNode *a, PQNode *b);
void printPQ(PriorityQueue *pq);
void freePQ(PriorityQueue *pq);

PriorityQueue* createPQ(int capacity, int isMinHeap) {
    PriorityQueue *pq = (PriorityQueue *)malloc(sizeof(PriorityQueue));
    pq->heap = (PQNode *)malloc(capacity * sizeof(PQNode));
    pq->size = 0;
    pq->capacity = capacity;
    pq->isMinHeap = isMinHeap;
    return pq;
}

void insertPQ(PriorityQueue *pq, int priority, int value) {
    if (pq->size == pq->capacity) {
        printf("Priority Queue is full\n");
        return;
    }
    PQNode node;
    node.priority = priority;
    node.value = value;

    pq->heap[pq->size] = node;
    heapifyUp(pq, pq->size);
    pq->size++;
}

int extractPQ(PriorityQueue *pq) {
    if (pq->size == 0) {
        printf("Priority Queue is empty\n");
        return -1;
    }
    
    int highestPriorityValue = pq->heap[0].value;  // get value
    pq->heap[0] = pq->heap[pq->size - 1];  // last element -> root
    pq->size--;
    heapifyDown(pq, 0);

    return highestPriorityValue;
}

// .. restore heap after insertion
void heapifyUp(PriorityQueue *pq, int index) {
    int parent = (index - 1) / 2;

    if (index > 0) {
        int condition = pq->isMinHeap
                        ? pq->heap[index].priority < pq->heap[parent].priority
                        : pq->heap[index].priority > pq->heap[parent].priority;

        if (condition) {
            swap(&pq->heap[index], &pq->heap[parent]);
            heapifyUp(pq, parent);
        }
    }
}

// .. restore heap after extraction
void heapifyDown(PriorityQueue *pq, int index) {
    int left = 2 * index + 1;
    int right = 2 * index + 2;
    int selected = index;

    if (pq->isMinHeap) {
        if (left < pq->size && pq->heap[left].priority < pq->heap[selected].priority)
            selected = left;
        if (right < pq->size && pq->heap[right].priority < pq->heap[selected].priority)
            selected = right;
    } else {
        if (left < pq->size && pq->heap[left].priority > pq->heap[selected].priority)
            selected = left;
        if (right < pq->size && pq->heap[right].priority > pq->heap[selected].priority)
            selected = right;
    }

    if (selected != index) {
        swap(&pq->heap[index], &pq->heap[selected]);
        heapifyDown(pq, selected);
    }
}

void swap(PQNode *a, PQNode *b) {
    PQNode temp = *a;
    *a = *b;
    *b = temp;
}


void printPQ(PriorityQueue *pq) {
    for (int i = 0; i < pq->size; i++)
        printf("(%d, %d) ", pq->heap[i].priority, pq->heap[i].value);
    printf("\n");
}


void freePQ(PriorityQueue *pq) {
    free(pq->heap);
    free(pq);
}


int main() {
    printf("Min-Priority Queue (lower number = higher priority):\n");
    PriorityQueue *minPQ = createPQ(10, 1);  // Min-Heap Priority Queue

    insertPQ(minPQ, 3, 100);
    insertPQ(minPQ, 1, 200);
    insertPQ(minPQ, 2, 300);

    printf("Queue: ");
    printPQ(minPQ);  // Output: [(1, 200), (3, 100), (2, 300)]
    
    printf("Extracted: %d\n", extractPQ(minPQ));  // Output: 200 (priority 1)
    printf("Queue after extraction: ");
    printPQ(minPQ);  // Output: [(2, 300), (3, 100)]
    
    printf("\nMax-Priority Queue (higher number = higher priority):\n");
    PriorityQueue *maxPQ = createPQ(10, 0);  // Max-Heap Priority Queue

    insertPQ(maxPQ, 3, 100);
    insertPQ(maxPQ, 1, 200);
    insertPQ(maxPQ, 2, 300);

    printf("Queue: ");
    printPQ(maxPQ);  // Output: [(3, 100), (1, 200), (2, 300)]
    
    printf("Extracted: %d\n", extractPQ(maxPQ));  // Output: 100 (priority 3)
    printf("Queue after extraction: ");
    printPQ(maxPQ);  // Output: [(2, 300), (1, 200)]

    freePQ(minPQ);
    freePQ(maxPQ);

    return 0;
}

File: workbook/ch07/data/common/pqueue.py

class PriorityQueue:
    def __init__(self, is_min_heap=True):
        self.heap = []
        self.is_min_heap = is_min_heap

    def _parent(self, index):
        return (index - 1) // 2

    def _left_child(self, index):
        return 2 * index + 1

    def _right_child(self, index):
        return 2 * index + 2

    def insert(self, priority, value):
        entry = (priority, value) if self.is_min_heap else (-priority, value)
        self.heap.append(entry)
        self._heapify_up(len(self.heap) - 1)

    def _heapify_up(self, index):
        parent = self._parent(index)
        if index > 0 and self.heap[index][0] < self.heap[parent][0]:  
            self.heap[index], self.heap[parent] = self.heap[parent], self.heap[index]
            self._heapify_up(parent)

    def extract(self):
        if not self.heap:
            raise IndexError("Priority queue is empty")
        if len(self.heap) == 1:
            return self.heap.pop()[1]  # only value

        root_value = self.heap[0][1]  # extract value
        self.heap[0] = self.heap.pop()  # last element -> root
        self._heapify_down(0)
        return root_value

    def _heapify_down(self, index):
        left = self._left_child(index)
        right = self._right_child(index)
        smallest = index

        if left < len(self.heap) and self.heap[left][0] < self.heap[smallest][0]:
            smallest = left
        if right < len(self.heap) and self.heap[right][0] < self.heap[smallest][0]:
            smallest = right

        if smallest != index:
            self.heap[index], self.heap[smallest] = self.heap[smallest], self.heap[index]
            self._heapify_down(smallest)

    def peek(self):
        if not self.heap:
            raise IndexError("Priority queue is empty")
        return self.heap[0][1]

    def is_empty(self):
        return len(self.heap) == 0

    def __str__(self):
        return str([(p if self.is_min_heap else -p, v) for p, v in self.heap])
    

# example
print("Min-Priority Queue (lower number = higher priority):")
min_pq = PriorityQueue(is_min_heap=True)
min_pq.insert(3, "Task A")
min_pq.insert(1, "Task B")
min_pq.insert(2, "Task C")

print("Queue:", min_pq)  # Output: [(1, 'Task B'), (3, 'Task A'), (2, 'Task C')]
print("Extracted:", min_pq.extract())  # Output: 'Task B' (priority 1)
print("Queue after extraction:", min_pq)  # Output: [(2, 'Task C'), (3, 'Task A')]

print("\nMax-Priority Queue (higher number = higher priority):")
max_pq = PriorityQueue(is_min_heap=False)
max_pq.insert(3, "Task A")
max_pq.insert(1, "Task B")
max_pq.insert(2, "Task C")

print("Queue:", max_pq)  # Output: [(3, 'Task A'), (1, 'Task B'), (2, 'Task C')]
print("Extracted:", max_pq.extract())  # Output: 'Task A' (priority 3)
print("Queue after extraction:", max_pq)  # Output: [(2, 'Task C'), (1, 'Task B')]
